### PR TITLE
Jira issue push improvements

### PR DESCRIPTION
**Pull request by chillu**
Followup from https://github.com/vipsoft/JiraExtension/pull/17

### CODE REVIEW

**Comment by chillu** [comment]
@srizzling Can you please keep an eye on this?

**Comment by srizzling** [comment]
Yep. Will do this sometime this week. Probably requires a bit of cleanup from my end in terms of documentation as well.

**Comment by robocoder** [comment]
Ok, ping me when you're done.  This extension is on my todo list for Behat3 compatibility.

**Comment by srizzling** [comment]
@robocoder Whats your convention on commits? Should I squash these all to one commit or would you want separated commits for different features?

**Comment by robocoder** [comment]
I prefer squashing when subsequent commits revert/undo previous commits.  Otherwise, I'm ok with either squashing or separate commits.

**Comment by srizzling** [comment]
@robocoder  I have mostly refactored the code add some tests as well but would need some peer review on the different commits. /cc @chillu 

Will still need to add some documentation, documenting the different parameters in the behat.yml file. Would be good if you could have a look over it. 

12fb819 - Fixes a bug where if there no backgrounds on the issue. Also adds a little formatting to be viewed on Jira (using the {code} tags used to "load" the jira ticket"

a4e46d1 - Adds some (not alot) information once the pushing of issues was completed

16269df - Disallows updating of tasks that are in the ignored_issues parameter which prevents CI from pushing to different status that are open. 

Will plan to finish writing the documentation by the weekend. :)

**Comment by srizzling** [comment]
I have added the documentation for using plugin. If you require more information I can add it.

File: doc/index.rst [hunk 1]
```diff
@@ -2,7 +2,8 @@
 JiraExtension
 =============
 
-The Jira extension allows features to be loaded from Jira issues.
+The Jira extension allows features to be pushed and loaded from Jira issues. Pushing, and loading issues are two seperate features of the JiraExtension.
```
Comment (robocoder) [inline doc/index.rst]: Sorry, grammar nazi mode.

> Pushing and loading issues are two separate features of the JiraExtension.

File: doc/index.rst [hunk 2]
```diff
@@ -42,6 +43,8 @@ Configuration
 =============
 Activate extension in your **behat.yml** and define your Jira connection settings:
 
+**Note:** Its improtant for the user, you use the username not the email address.
```
Comment (robocoder) [inline doc/index.rst]: > It's important for the `user` that you use the username not the email address.

File: doc/index.rst [hunk 3]
```diff
@@ -77,6 +80,29 @@ You can be more specific in your search if you create a custom issue type, and u
...
+        VIPSoft\JiraExtension\Extension:
+          host: http://jira.example.com:8080/ # Host Address of the Jira Application
+          user: jirauser # Username for the SOAP API (Use username not email)
+          password: secret # Password for the SOAP API
+          jql: "summary ~ 'Feature'" # JQL to locate a custom issue type
+          comment_on_pass: false # Comment on the issue if it passes
+          comment_on_fail: true # Comment on the issue if it fails
+          reopen_on_fail: false # Reopen if it fails
+          cache_directory: /tmp/behat-jira 
+          push_issue: false # Create behat tests, that "pushed" to the assoicated Jira Ticket
```
Comment (robocoder) [inline doc/index.rst]: > Run behat tests and "push" to the associated Jira Ticket

File: doc/index.rst [hunk 4]
```diff
@@ -77,6 +80,29 @@ You can be more specific in your search if you create a custom issue type, and u
...
+          ignored_statuses: "Closed,Resolved" # Comma seperated list of ignored jira updates. Issues with these statues, will not be pushed to Jira.
```
Comment (robocoder) [inline doc/index.rst]: > Comma separated list of ignored Jira updates. Issues with these statuses will not be pushed to Jira.

File: doc/index.rst [hunk 5]
```diff
@@ -77,6 +80,29 @@ You can be more specific in your search if you create a custom issue type, and u
...
+          user: jirauser # Username for the SOAP API (Use username not email)
+          password: secret # Password for the SOAP API
+          jql: "summary ~ 'Feature'" # JQL to locate a custom issue type
+          comment_on_pass: false # Comment on the issue if it passes
+          comment_on_fail: true # Comment on the issue if it fails
+          reopen_on_fail: false # Reopen if it fails
+          cache_directory: /tmp/behat-jira 
+          push_issue: false # Create behat tests, that "pushed" to the assoicated Jira Ticket
+          ignored_statuses: "Closed,Resolved" # Comma seperated list of ignored jira updates. Issues with these statues, will not be pushed to Jira. 
+          tag_pattern: /jira:(.*)/ # A pattern to match the revelat tags to push to jira
```
Comment (robocoder) [inline doc/index.rst]: > A pattern to match the revelant tags to push to Jira

File: doc/index.rst [hunk 6]
```diff
@@ -130,6 +159,32 @@ Usage
...
+        Feature: Jira integration
+          In order to facilitate the authoring of Behat features by non-developers
+          As a developer
+          I want to write an extension to load features from Jira issues.
+
+        Scenario: Push Me!
+            Given I am a Jira issue
+            And I contain a Behat feature
+            When I am loaded by JiraExtension
+            Then I should parsed by Gherkin
```
Comment (robocoder) [inline doc/index.rst]: > Then I should be parsed  by Gherkin

File: doc/index.rst [hunk 7]
```diff
@@ -130,6 +159,32 @@ Usage
...
+            And I should be pushed to Jira
+    {code}
+
+3. Given a issue has been created already, use the id (for example: JRA-1) to create a tag on the scenario. If you are using the default tag pattern, it will be like this: @jira:JRA-1.
+
+4. Running bin/behat --name="Push Me!" should push the issue to Jira and to the assoicated ticket.
```
Comment (robocoder) [inline doc/index.rst]: > associated

File: src/VIPSoft/JiraExtension/EventListener/HookListener.php [hunk 8]
```diff
@@ -154,8 +158,10 @@ public function parseJiraTags($tags)
      * @return Array $jiraTags
      */
     public function getStepText($scenario)
-    { 
+    {
+
```
Comment (robocoder) [inline src/VIPSoft/JiraExtension/EventListener/HookListener.php:161]: This is shorter:

```
$stepArray = array("{code}");
```

File: src/VIPSoft/JiraExtension/EventListener/HookListener.php [hunk 9]
```diff
@@ -167,13 +173,11 @@ public function getStepText($scenario)
...
             $backgroundSteps = $background -> getSteps();
-
-            if ($backgroundSteps -> hasSteps()){
-                $stepArray[] = "Background: " . $background -> getTitle();
-            }
-
-            foreach ($scenarioSteps as $step) {
-                $stepArray[] = $step->getType()." ".$step->getText();
+     
+            $stepArray[] = "Background: " . $background -> getTitle();
```
Comment (robocoder) [inline src/VIPSoft/JiraExtension/EventListener/HookListener.php:174]: Please drop whitespace around `->`:

```
$stepArray[] = "Background: " . $background->getTitle();
```

File: src/VIPSoft/JiraExtension/Extension.php [hunk 10]
```diff
@@ -55,6 +55,9 @@ public function load(array $config, ContainerBuilder $container)
         if (isset($config['push_issue'])) {
             $container->setParameter('behat.jira.push_issue', $config['push_issue']);
         }
+        if(isset($config['ignored_statuses'])){
```
Comment (robocoder) [inline src/VIPSoft/JiraExtension/Extension.php]: Please add a space after `if`, and between `) {`.

File: src/VIPSoft/JiraExtension/Service/JiraService.php [hunk 11]
```diff
@@ -55,6 +55,20 @@ class JiraService
      */
     private $store;
 
+     /**
```
Comment (robocoder) [inline src/VIPSoft/JiraExtension/Service/JiraService.php]: Please fix alignment of comment block.

File: src/VIPSoft/JiraExtension/Service/JiraService.php [hunk 12]
```diff
@@ -55,6 +55,20 @@ class JiraService
...
+    private $pushedIssue;
+
+
+    /**
+     * @var string $ignoredStatus
+     */
+    private $ignoredStatus;
+
+
+    private $ignoredIssues;
```
Comment (robocoder) [inline src/VIPSoft/JiraExtension/Service/JiraService.php:71]: Missing docblock.

```
/**
 * @var array
 */
```

File: src/VIPSoft/JiraExtension/Service/JiraService.php [hunk 13]
```diff
@@ -64,8 +78,9 @@ class JiraService
      * @param string      $password     Jira user password
      * @param string      $jql          JQL query
      * @param string      $featureField Jira field to upload feature file
+     * @param string       $ignoredStatus Tickets with the statuses in this array are ignored
```
Comment (robocoder) [inline src/VIPSoft/JiraExtension/Service/JiraService.php]: Please fix alignment of parameters in docblock.

File: src/VIPSoft/JiraExtension/Service/JiraService.php [hunk 14]
```diff
@@ -271,12 +291,35 @@ public function compareIssueField($jiraTicket, $value){
...
-        }
+        } 
 
         return true; //Issues are not the same
     }
 
     /**
+     * This method checks the status of the Jira Ticket to ensure that
+     * it is not in the array of ignored statuses. 
+     *
```
Comment (robocoder) [inline src/VIPSoft/JiraExtension/Service/JiraService.php:306]: Missing @param in docblock.

File: src/VIPSoft/JiraExtension/Service/JiraService.php [hunk 15]
```diff
@@ -271,12 +291,35 @@ public function compareIssueField($jiraTicket, $value){
...
+        } 
 
         return true; //Issues are not the same
     }
 
     /**
+     * This method checks the status of the Jira Ticket to ensure that
+     * it is not in the array of ignored statuses. 
+     * 
+     * @return array jiraticket to check
```
Comment (robocoder) [inline src/VIPSoft/JiraExtension/Service/JiraService.php]: Return type is now `boolean`.

File: src/VIPSoft/JiraExtension/Service/JiraService.php [hunk 16]
```diff
@@ -271,12 +291,35 @@ public function compareIssueField($jiraTicket, $value){
...
+     */
+    public function checkStatus($ticket)
```
Comment (robocoder) [inline src/VIPSoft/JiraExtension/Service/JiraService.php]: Maybe a better name for this method would be `isIgnored()` and inverting the return value?

**Comment by robocoder** [comment]
Otherwise it looks good.  If you don't get around to making the suggested changes, I can tackle it after I merge.

**Comment by srizzling** [comment]
I'll do these changes. :) Thanks for that!

**Comment by srizzling** [comment]
@robocoder  I have implemented most of your suggestions and I ran phpcs using the Symfony standard to find any whitespace and other coding conventions I have forgotten.  I'm happy to squash if you would like.

**Comment by robocoder** [comment]
Feel free to squash

**Comment by srizzling** [comment]
Can I squash everything into one commit?

**Comment by srizzling** [comment]
Done anyway :)

**Comment by robocoder** [comment]
Thanks.

**Pull request merged**